internal/db/index.go

```diff
@@ -430,3 +425,323 @@ func isUpdatingIndexedFields(index CollectionIndex, oldDoc, newDoc *client.Docum
 	}
 	return false
 }
+
+type collectionArrayBaseIndex struct {
+	collectionBaseIndex
+	arrFieldsIndexes []int
+}
+
+func newCollectionArrayBaseIndex(base collectionBaseIndex) collectionArrayBaseIndex {
+	ind := collectionArrayBaseIndex{collectionBaseIndex: base}
+	for i := range base.fieldsDescs {
+		if base.fieldsDescs[i].Kind.IsArray() {
+			ind.arrFieldsIndexes = append(ind.arrFieldsIndexes, i)
+		}
+	}
+	if len(ind.arrFieldsIndexes) == 0 {
+		return collectionArrayBaseIndex{}
+	}
+	return ind
+}
+
+// newIndexKeyGenerator creates a function that generates index keys for a document
+// with multiple array fields.
+// All generated keys are unique.
+// For example for a doc with these values {{"a", "b", "a"}, {"c", "d", "e"}, {"f", "g"}} it generates:
+// "acf", "acg", "adf", "adg", "aef", "aeg", "bcf", "bcg", "bdf", "bdg", "bef", "beg"
+// Note: the example is simplified and doesn't include field separation
+func (index *collectionArrayBaseIndex) newIndexKeyGenerator(
+	doc *client.Document,
+	appendDocID bool,
+) (func() (core.IndexDataStoreKey, bool), error) {
+	key, err := index.getDocumentsIndexKey(doc, appendDocID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Collect unique values to use as source for generating keys
+	normValsArr := make([][]client.NormalValue, 0, len(index.arrFieldsIndexes))
+	for _, arrFieldIndex := range index.arrFieldsIndexes {
+		arrVal := key.Fields[arrFieldIndex].Value
+		normVals, err := client.ToArrayOfNormalValues(arrVal)
+		if err != nil {
+			return nil, err
+		}
+		sets := make(map[client.NormalValue]struct{})
+		for i := len(normVals) - 1; i >= 0; i-- {
+			if _, ok := sets[normVals[i]]; ok {
+				normVals[i] = normVals[len(normVals)-1]
+				normVals = normVals[:len(normVals)-1]
+			} else {
+				sets[normVals[i]] = struct{}{}
+			}
+		}
+		normValsArr = append(normValsArr, normVals)
+	}
+
+	arrFieldCounter := make([]int, len(index.arrFieldsIndexes))
+	done := false
+
+	return func() (core.IndexDataStoreKey, bool) {
+		if done {
+			return core.IndexDataStoreKey{}, false
+		}
+
+		resultKey := core.IndexDataStoreKey{
+			CollectionID: key.CollectionID,
+			IndexID:      key.IndexID,
+			Fields:       make([]core.IndexedField, len(key.Fields)),
+		}
+		copy(resultKey.Fields, key.Fields)
+
+		for i, counter := range arrFieldCounter {
+			field := &resultKey.Fields[index.arrFieldsIndexes[i]]
+			field.Value = normValsArr[i][counter]
+		}
+
+		for i := len(arrFieldCounter) - 1; i >= 0; i-- {
+			arrFieldCounter[i]++
+			if arrFieldCounter[i] < len(normValsArr[i]) {
+				break
+			}
+			arrFieldCounter[i] = 0
+			if i == 0 {
+				done = true
```

Comment:
todo: Please explain why this is code does this

Reply:
done

Reply:
okay got it - thanks Islam, makes sense to me now :)